[WIP] Add help text on how to interact with message links.

Sushmey opened the pull request:
<!-- See README for documentation, or ask in #zulip-terminal if unclear -->
### What does this PR do, and why?

### External discussion & connections
<!-- [x] all that apply, specifying topic and adding numbers after # for issues/PRs -->
- [x] Discussed in [Open external links with browser #T1440](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Open.20external.20links.20with.20browser.20.23T1440)
- [ ] Fully fixes #
- [x] Partially fixes issue #1372 
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?
<!-- [x] all that apply -->
- [x] Manually - Behavioral changes
- [x] Manually - Visual changes
- [ ] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [ ] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [ ] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [ ] It has a commit summary describing the  motivation and reasoning for the change
- [ ] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [ ] It flows clearly from a previous branch commit, and/or prepares for the next commit

### Visual changes    <!-- DELETE SECTION IF NO VISUAL CHANGE -->
<!-- Zulip tips at https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html -->
<!-- For video, try asciinema; after uploading, embed using
[![yourtitle](https://asciinema.org/a/<id>.png)](https://asciinema.org/a/<id>)
-->
<!-- NOTE: Attached videos/images will be clearer from smaller terminal windows -->
<img width="599" alt="Screenshot 2024-03-28 at 11 38 11 PM" src="https://github.com/zulip/zulip-terminal/assets/71403193/f0128b9b-bedd-42d7-ac9b-37a70145dfab">

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1605,7 +1621,21 @@ def __init__(
+            msg_info.append(
+                (
+                    (
+                        "Message Links",
+                        f"Press {interact_with_link_keys} to interact with links",
```
This change is distinct from the change to the category feature that supports it - or to put it differently: can this specific change be included only with similar changes for the other categories which may appear?

This feature can also be applied to other popups, though we may want to discuss that first.

neiljp replied:
I'm not sure if we discussed this in the stream, but I was referring to having this in a different commit to the one which adds category heading support.

neiljp reviewed (commented):
@Sushmey I left some preliminary feedback. Please let me know here or on czo if you have points that I've not answered.

Sushmey committed:
views: Comment this to resolve issue #1226.

This is not a part of this PR. This is just to resolve issue #1226.

Sushmey committed:
tests: ui_tools: Add test for `make_table_with_categories`.

Sushmey commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1041,16 +1041,32 @@ def make_table_with_categories(
                 if len(widgets) > 0:  # Separate categories with newline.
                     widgets.append(urwid.Text(""))
-                widgets.append(urwid.Text(("popup_category", category)))
+                if isinstance(category, tuple):
```
Yes ran into this while pushing but couldn't figure it out how to resolve it. Now I added the new case in the PopUpViewTableContent type so no more error :)

Sushmey commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1041,16 +1041,32 @@ def make_table_with_categories(
+                            [(column_widths[0], urwid.Text(label)), urwid.Text(data)],
+                            dividechars=dividechars,
+                        )
                     widgets.append(
                         urwid.AttrWrap(strip, None if index % 2 else "popup_contrast")
                     )
```
Yes updating the loop index is confusing. Figured out a better way to get what I expected.

No we don't need to wrap the popup_contrast. I did it because I was trying to make it visually discernible, updated it to work without the wrap.

We can have a discussion about the style.

neiljp replied:
We don't necessarily need to think about a style in this first version, but at the time it struck me as an additional motivator for adjusting the loop.

neiljp commented:
I believe I covered your questions so far in the stream or here.

neiljp commented on tests/ui_tools/test_popups.py:
```diff
@@ -179,6 +180,44 @@ def test_keypress_navigation(
+                        ],
+                    )
+                ],
+                [21, 24],
+                [Text, AttrWrap, AttrWrap, AttrWrap],
+            ),
+            ([("", "")], [21, 24], []),
```
The `21` and `24` are the same in each test you have in this PR. Are they constants which must work with the test data in particular, or can these be constants in the test parameters/body, and simplify the test inputs?

Sushmey replied:
No they are not constants, I had added them as constants for temporary purposes just to test if the test worked. Changed it so that it's not hardcoded now.

neiljp commented on tests/ui_tools/test_popups.py:
```diff
@@ -179,6 +180,44 @@ def test_keypress_navigation(
+        [
+            "contents",
+            "column_widths",
+            "expected_instance",
```
`instance` is rather vague - what does this function return?

I believe each part of the `instance` is essentially a row, which we don't mention in the docstring and maybe in the code that uses it, but could be a useful refactor?

Sushmey replied:
Changed the parameter name to `expected_widget_type` to be more descriptive since the docstring also mentions that the function is returning list of widgets.

Re second part, are you saying each widget of the list is essentially a row and that the docstring needs a refactor? I feel the current docstring is good enough, it does return a list of widgets that is used to make the table.

neiljp replied:
Using `widget` is consistent with the current docstring and terms used :+1:

The second point was my noting that the code might be more readable if we referred to the returned data as rows; widgets are any UI element, so are less specific.

neiljp commented on tests/ui_tools/test_popups.py:
```diff
@@ -179,6 +180,44 @@ def test_keypress_navigation(
+            "widget_with_str",
+        ],
+    )
+    def test_make_table_with_categories(
```
Having this test is great, but it does focus on the urwid form of the output, so it may be worth adding a suffix to the test name to emphasize that, eg. `__instances` based on your naming elsewhere (which can be improved).

Sushmey replied:
Adding an `__instances` suffix. Although I wonder if this implies there's a test that focuses on `make_table_with_categories` as a whole and not on some specific thing like instances.

neiljp replied:
I agree that it could suggest another test exists, but also means it won't conflict with another later ;) If there are multiple tests for a function, my preferred form now is to separate the function and the intention behind the test.

I suggested `__instances` since that's what you were checking in the test, and used in the expected part of the test case name, as per another comment - so let's sync them up if you think the updated expected parameter name is better :)

Sushmey replied:
I'e used `types` everywhere instead of instance because it seems more intuitive than `instances`. So let's change it to `__types` in the function name also?

neiljp replied:
Yeah, that'd be good :+1:

neiljp reviewed (commented):
@Sushmey It'd be great to get the category headings infrastructure merged, so I left some more feedback on what you have here :+1:

Please do reply if you have other issues with the typing.

neiljp commented:
Should the linked issue be #1372, not #1002?

Sushmey committed:
views: Amend the function to add help text.

Updating the function `make_table_with_categories`.

So that right aligned hint text can be appended to heading.

Sushmey committed:
views: Add help text on how to interact with message links.

Add help text.

neiljp reviewed (commented):
@Sushmey Glad to see you got the types sorted :+1: There seems to be a failing test? Maybe try `check-branch` locally, to see what commit is failing?

I replied to some previous comments and left a few more inline.

The PR could do with removing the exploratory parts, eg. commented out, and that would make it clearer what exactly is changing.

Sushmey committed:
views: Amend function to calculate width with help text.

Amend `calculate_table_widths` to add case of category with help text.

Sushmey committed:
tests: ui_tools: Add test for `calculate_table_width`.

Sushmey commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1603,7 +1611,7 @@ def __init__(
             ", ".join(map(str, display_keys_for_command("FULL_RAW_MESSAGE")))
         )
-        msg_info = [
+        msg_info: PopUpViewTableContent = [
```
It doesn't work otherwise since `PopUpViewTableContent` also has types which doesn't support append so need an explicit assertion. 
I can't not explicitly type it as `PopUpViewTableContent` since Mypy assumes it to be the type of the first data assigned to it.

Sushmey commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1042,19 +1050,35 @@ def make_table_with_categories(
                 if len(widgets) > 0:  # Separate categories with newline.
                     widgets.append(urwid.Text(""))
-                widgets.append(urwid.Text(("popup_category", category)))
+                if isinstance(category, tuple) and isinstance(content, list):
+                    content.append(category)
```
Thanks for pointing it out. There was no reason to put it as content. Changed it and removed the condition too.

Sushmey commented on tests/ui_tools/test_popups.py:
```diff
@@ -179,6 +180,44 @@ def test_keypress_navigation(
+        popup_width, column_widths = self.pop_up_view.calculate_table_widths(
+            contents, len(self.title)
+        )
+        widgets = self.pop_up_view.make_table_with_categories(contents, column_widths)
+        for index, widget in enumerate(widgets):
+            assert isinstance(widget, expected_widget_type[index])
```
zip is used to combine two iterables, right? Not sure how we can use that here.

neiljp replied:
Yep, I already saw you changed it in the code :)
